### PR TITLE
#32: clients detection

### LINKED ISSUE #32: Client detection
Add Client detection UI component. [Description](https://github.com/user-attachments/files/16993918/Client.Detection.pdf) provided.

### PR DESCRIPTION
**Pull request by shurshurik**
Add a button to detect new clients

### CODE REVIEW

File: vpn-router-bootstrap/build.gradle.kts [hunk 1]
```diff
@@ -3,6 +3,8 @@ dependencies {
     implementation(project(":vpn-router-api"))
     implementation(project(":vpn-router-web"))
     implementation(project(":vpn-router-alice"))
+    implementation("com.vaadin:vaadin-spring-boot-starter:24.4.8")
```
Comment (KirillSmirnow) [inline vpn-router-bootstrap/build.gradle.kts]: No, bootstrap may not depend on web

File: vpn-router-bootstrap/src/main/java/vpnrouter/VpnRouter.java [hunk 2]
```diff
@@ -1,10 +1,16 @@
...
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
+import org.springframework.scheduling.annotation.EnableAsync;
 
 @SpringBootApplication
-public class VpnRouter {
+@Push
+@EnableAsync
+public class VpnRouter extends SpringBootServletInitializer implements AppShellConfigurator {
```
Comment (KirillSmirnow) [inline vpn-router-bootstrap/src/main/java/vpnrouter/VpnRouter.java]: No, bootstrap may not depend on web

File: vpn-router-api/src/main/java/vpnrouter/api/client/ClientDetectionService.java [hunk 3]
```diff
@@ -5,6 +5,7 @@ public interface ClientDetectionService {
     void detectAndSave(CompletionListener completionListener);
 
     interface CompletionListener {
+        void onStart();
```
Comment (KirillSmirnow) [inline vpn-router-api/src/main/java/vpnrouter/api/client/ClientDetectionService.java:8]: No, you've broken the logic: this is a completion listener and onStart is not completion.

We may redesign the behavior.

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/ClientDetectionButtonFactory.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+@RequiredArgsConstructor
+public class ClientDetectionButtonFactory {
+    private final ClientDetectionService clientDetectionService;
+
+    public Button build() {
+        var detectionButton = new Button("Detect clients");
+        detectionButton.addClickListener(event -> {
+                    UI ui = event.getSource().getUI().orElseThrow();
+                    clientDetectionService.detectAndSave(new CompetitionListenerImpl(ui));
+                }
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/ClientDetectionButtonFactory.java:20]: Brackets should open and close together on same correspondent lines like this:
```
[<((  // opening 1
({({  // opening 2

// some code

})})  // closing 2
))>]  // closing 1
```

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/ClientDetectionButtonFactory.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+@Component
+@RequiredArgsConstructor
+public class ClientDetectionButtonFactory {
+    private final ClientDetectionService clientDetectionService;
+
+    public Button build() {
+        var detectionButton = new Button("Detect clients");
+        detectionButton.addClickListener(event -> {
+                    UI ui = event.getSource().getUI().orElseThrow();
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/ClientDetectionButtonFactory.java:18]:
```suggestion
                    var ui = UI.getCurrent();
```

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/ClientsPage.java [hunk 6]
```diff
@@ -36,4 +39,8 @@ private Button buildAddClientButton() {
                 event -> getUI().ifPresent(ui -> ui.navigate(AddClientPage.class))
         );
     }
+
+    private Button buildDetectClientButton() {
+        return clientDetectionButtonFactory.build();
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/ClientsPage.java:44]: The wrapping method is redundant

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import com.vaadin.flow.component.notification.Notification;
+import lombok.RequiredArgsConstructor;
+import vpnrouter.api.client.ClientDetectionService;
+
+import java.util.Timer;
+import java.util.TimerTask;
+
+@RequiredArgsConstructor
+public class CompetitionListenerImpl implements ClientDetectionService.CompletionListener {
+    private final int DELAY_MILLISECONDS = 1000;
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java:13]: Let's go somewhere and learn how to declare constants in Java

Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java:13]: Why kind of delay is this (should be specified)?

Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java:13]:
```suggestion
    private final int DELAY_IN_MILLISECONDS = 1000;
```

Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java:13]: better: `Duration myDelay = Duration.ofSeconds(1)`

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import com.vaadin.flow.component.UI;
+import com.vaadin.flow.component.notification.Notification;
+import lombok.RequiredArgsConstructor;
+import vpnrouter.api.client.ClientDetectionService;
+
+import java.util.Timer;
+import java.util.TimerTask;
+
+@RequiredArgsConstructor
+public class CompetitionListenerImpl implements ClientDetectionService.CompletionListener {
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java:12]: 🐶 🐕 🦮

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Override
+    public void onNewClientsNotFound() {
+        ui.access(() -> Notification.show("Detection completed: no new clients found"));
+    }
+
+    @Override
+    public void onNewClientsFound(int newClientsCount) {
+        ui.access(() -> {
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java:28]: It seems code format is poor here

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                    Notification.show(
+                            "Detection completed: %s new clients found".formatted(newClientsCount),
+                            DELAY_MILLISECONDS,
+                            Notification.Position.BOTTOM_START
+                    );
+                    new Timer().schedule(
+                            new TimerTask() {
+                                @Override
+                                public void run() {
+                                    ui.access(() -> ui.getPage().reload());
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java:38]: Page reloading is too rough method to refresh data -- should be a bit smarter/optimal

**Review by KirillSmirnow — CHANGES_REQUESTED** [review]

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/ClientsPage.java [hunk 11]
```diff
@@ -34,10 +35,9 @@ public void onAttach(AttachEvent event) {
     }
 
     private Button buildAddClientButton() {
-        return new Button(
-                "Add client",
-                event -> getUI().ifPresent(ui -> ui.navigate(AddClientPage.class))
-        );
+        var button = new Button(VaadinIcon.PLUS.create());
```
Comment (KirillSmirnow) [inline vpn-router-web/src/main/java/vpnrouter/web/ui/clients/ClientsPage.java:38]: 👍

File: vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java (hunk 8 above)
Comment (shurshurik) [reply vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java:12]: What do you mean?))

**Pull request closed without merging**